extensions/amp-auto-lightbox/0.1/amp-auto-lightbox.js

```diff
@@ -0,0 +1,336 @@
+  'BlogPosting',
+  'LiveBlogPosting',
+  'DiscussionForumPosting',
+];
+
+/** Factor of naturalArea vs renderArea to lightbox. */
+export const RENDER_AREA_RATIO = 1.2;
+
+/** Factor of renderArea vs viewportArea to lightbox. */
+export const VIEWPORT_AREA_RATIO = 0.3;
+
+const ACTIONABLE_ANCESTORS =
+    'a[href], amp-selector > [option], amp-script, amp-story';
+
+
+/** @visibleForTesting */
+export class Criteria {
+
+  /**
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  static meetsAll(element) {
+    return Criteria.meetsPlaceholderCriteria(element) &&
+      Criteria.meetsSizingCriteria(element) &&
+      Criteria.meetsActionableCriteria(element);
+  }
+
+  /**
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  static meetsSizingCriteria(element) {
+    const {naturalWidth, naturalHeight} =
+        dev().assertElement(element.querySelector('img'));
+
+    const {width: renderWidth, height: renderHeight} = element.getLayoutBox();
+
+    const viewport = Services.viewportForDoc(element);
+    const {width: vw, height: vh} = viewport.getSize();
+
+    return meetsSizingCriteria(
+        renderWidth,
+        renderHeight,
+        naturalWidth,
+        naturalHeight,
+        vw,
+        vh);
+  }
+
+  /**
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  static meetsPlaceholderCriteria(element) {
+    return !closestBySelector(element, '[placeholder]');
+  }
+
+  /**
+   * @param {!Element} element
+   * @return {boolean}
+   */
+  static meetsActionableCriteria(element) {
+    if (closestBySelector(element, ACTIONABLE_ANCESTORS)) {
+      return false;
+    }
+    const actions = Services.actionServiceForDoc(element);
+    return !actions.hasAction(element, 'tap');
+  }
+}
+
+
+/**
+ * @param {number} renderWidth
+ * @param {number} renderHeight
+ * @param {number} naturalWidth
+ * @param {number} naturalHeight
+ * @param {number} vw
+ * @param {number} vh
+ * @return {boolean}
+ * @visibleForTesting
+ */
+export function meetsSizingCriteria(
+  renderWidth,
+  renderHeight,
+  naturalWidth,
+  naturalHeight,
+  vw,
+  vh) {
+
+  const viewportArea = vw * vh;
+  const naturalArea = naturalWidth * naturalHeight;
+  const renderArea = renderWidth * renderHeight;
+
+  const isShrunk =
+    (naturalArea / renderArea) >= RENDER_AREA_RATIO;
+
+  const isCoveringSignificantArea =
+    (renderArea / viewportArea) >= VIEWPORT_AREA_RATIO;
+
+  const isLargerThanViewport = naturalWidth > vw || naturalHeight > vh;
+
+  return isShrunk || isLargerThanViewport || isCoveringSignificantArea;
+}
+
+
+/** @visibleForTesting */
+export class Scanner {
+
+  /**
+   * @param {!Document} doc
+   * @return {<!Array<!Promise<!Element>>}
+   */
+  static getAllImages(doc) {
+    const imgs = toArray(doc.querySelectorAll('amp-img')).filter(img =>
+      !img.hasAttribute(LIGHTBOXABLE_ATTR) &&
+      !img.hasAttribute(VISITED_ATTR));
+
+    imgs.forEach(img => {
+      img.setAttribute(VISITED_ATTR, '');
```

Comment:
`VISITED_ATTR` is a public attribute (`data-amp-auto-lightbox-visited`), so it must be done in a mutate phase.